Tutorial: How to use React Charts to visualize analytics data (with examples) (#9245)
PostHog/posthog.com: contents/tutorials/react-charts.md

## Creating a Next.js app

To create a Next.js app, run the following command, choose all the default options, and select the `app` router.

```bash
npx create-next-app@latest react-charts-example
```

This creates a `react-charts-example` folder with everything we need to get started.

## Querying data from PostHog

Assuming you have data in PostHog to query, the next step is to set up our query request to PostHog.

> **Don't have data to query?** Check out [our guide to setting up analytics in Next.js](/tutorials/nextjs-app-directory-analytics) to start capturing some.

Start by creating a personal API key. You can do this by going to [personal API keys](https://us.posthog.com/settings/user-api-keys) in your project settings, clicking **Create personal API key**, giving it a label, choosing the **Performing analytics queries** scope (AKA query read), and clicking **Create key**.

<ProductScreenshot
  imageLight="https://res.cloudinary.com/dmukukwp6/image/upload/light_api_468581fdca.png" 
  imageDark="https://res.cloudinary.com/dmukukwp6/image/upload/dark_api_a755bfb6ac.png"
  alt="Creating a personal API key in PostHog" 
  classes="rounded"
/>

> > **⚠️ Warning:** The following is a simplified example. It will expose your private PostHog data to the public internet, so don't do this in real life. You want to make sure your personal API key isn't exposed and your query request happens securely on the server side.

Reviewer: This warning makes it seem like by creating an API key, all of our posthog data automatically becomes public. It also doesnt make sense here, since we havent yet added the code for the apikey on the client.

Edit: I dont actually see the part where we expose the API key? It's in the /api folder, which is server side, not client side? So we dont need this warning at all?

Author: If you implemented it this way, anyone could query the `api/query` endpoint with whatever HogQL payload they want and get whatever data they want. 

The right way to do it is to have the payload in the server side as well, but that would require changes to two files on every example which I didn't want to do for brevity.

Author: rewrote to clarify again

> > **⚠️ Warning:** The following is a simplified example. It will expose your private PostHog data to the public internet, so don't do this in real life. You want to make sure your personal API key isn't exposed and your query request happens securely on the server side.

Reviewer: ```suggestion
```

---
title: How to use React Charts to visualize analytics data (with examples)
date: 2024-08-23
author:
 - ian-vanagas
tags:
 - product analytics
---

> React Charts is a popular visualization and charting library for React. It provides a simplified set of performant charts to use with analytics data from PostHog.

Reviewer: ```suggestion
[React Charts](https://react-charts.tanstack.com/) is a popular visualization and charting library for React. It provides a simplified set of performant charts to use with analytics data from PostHog.
```

> To provide examples of what you can do with it, we create a basic Next.js app, set it up to get data from PostHog's query API, and visualize it with React Charts.

Reviewer: ```suggestion
To provide examples of what you can do with it, we create a basic Next.js app, set it up to fetch data from [PostHog's query API](/docs/api/query), and visualize it with React Charts.
```

> **Don't want to do all the work of querying and visualizing?** You can always share or embed insights, dashboards, and more. See [sharing docs](/docs/product-analytics/sharing) for more.

## Creating a Next.js app

> Because we need to query private data, we need a way to make server side requests. Next.js makes this easy (which is why we chose it).

Reviewer: I dont quite follow why this is?

Author: your PostHog data is private, querying it from the client side exposes your personal API key that could be used to query whatever data from PostHog you wanted.

Author: Yes, I put this warning below

> > **Note:** You'll likely want to include the payload on the server-side too so users can't make requests to this endpoint with arbitrary queries. We don't do this in this tutorial to keep the changes in the examples below in a single file.

Reviewer 2: I mean something more like 
> `!!!` **WARNING** `!!!` **This will expose all data in your PostHog account to the public internet** `!!!` **WARNING** `!!!` YOU DO NOT WANT TO DO THIS IN REAL LIFE

Author: 🫡

Reviewer 2: And then a followup tutorial: how to do this securely 😅

To create a Next.js app, run the following command, choose all the default options, and select the `app` router.

```bash
npx create-next-app@latest react-charts-example
```

This creates a `react-charts-example` folder with everything we need to get started.

## Querying data from PostHog

Assuming you have data in PostHog to query, the next step is to set up our query request to PostHog.

> > **Don't have data to query?** Check out [our guide to setting up analytics in Next.js](/tutorials/nextjs-app-directory-analytics) to start capturing some. Unlike other tutorials, we aren't going to install PostHog in this one.

Reviewer: Not needed IMO
```suggestion
> **Don't have data to query?** Check out [our guide to setting up analytics in Next.js](/tutorials/nextjs-app-directory-analytics) to start capturing some.
```

Start by creating a personal API key. You can do this by going to [personal API keys](https://us.posthog.com/settings/user-api-keys) in your project settings, clicking **Create personal API key**, giving it a label, choosing the **Performing analytics queries** scope (AKA query read), and clicking **Create key**.

<ProductScreenshot
  imageLight="https://res.cloudinary.com/dmukukwp6/image/upload/light_api_468581fdca.png" 
  imageDark="https://res.cloudinary.com/dmukukwp6/image/upload/dark_api_a755bfb6ac.png"
  alt="Creating a personal API key in PostHog" 
  classes="rounded"
/>

Next, go into your `react-charts-example` folder, then the `app` folder, and create an `api` folder. In the `api` folder, create a `query.js` file. Here we set up a fetch request to PostHog's [query API endpoint](/docs/api/query) using your project ID (from your PostHog URL) and personal API key like this:

> ```js
> // app/api/query.js

Reviewer: ```suggestion
```js file=app/api/query.js
```

export const fetchQuery = async (payload) => {

>   const projectId = "12345";

Reviewer: Would be helpful to point ou where this is exactly. I know you mention in the top, but helpful to be a bit more prescriptive
```suggestion
  const projectId = "12345"; // e.g. if your PostHog dashboard url is https://us.posthog.com/project/12345/

>   const apiKey = "phx_1a2b3c4d5e6f7g8h9i0j1k2l3m4n5o6p7q8r9s0tuvwxyz";

Reviewer: ```suggestion
  const apiKey = "<your_personal_api_key>";
```

  try {
    const response = await fetch(`https://us.posthog.com/api/projects/${projectId}/query/`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${apiKey}`
      },
      body: JSON.stringify(payload),
    });
    const data = await response.json();
    return data;
  } catch (error) {
    console.error("Error fetching data:", error);
    throw error;
  }
}
```

> > **Note:** You'll likely want to include the payload on the server-side too so users can't make requests to this endpoint with arbitrary queries. We don't do this in this tutorial to keep the changes in the examples below in a single file.

Reviewer: I dont understand this note. What do we mean by including the payload on the serverside? arent we in the server side here?

> Finally, in our client component, we make the query request using [HogQL](/docs/product-analytics/sql) and display the data:

Reviewer: It's worth mentioning here that you can also make requests without HogQL right? What are the other ways?

Reviewer: I think we should add a note that there are other ways to do it and not just HogQL

Author: Other options would require using [undocumented](https://posthog.com/docs/api/query) query types like `TrendsQuery` which is more confusing I think
